exercises/bowling/BowlingTest.cs

```diff
@@ -1,207 +1,350 @@
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        Assert.Equal(18, sut.Score());
     }
 
     [Fact(Skip = "Remove to run test")]
     public void Rolling_a_spare_with_the_two_roll_bonus_does_not_get_a_bonus_roll()
     {
-        var rolls = new[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 7, 3 };
-        var game = RollMany(rolls, new BowlingGame());
-        Assert.Equal(20, game.Score());
+        var previousRolls = new int[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 7, 3};
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        Assert.Equal(20, sut.Score());
     }
 
     [Fact(Skip = "Remove to run test")]
     public void Strikes_with_the_two_roll_bonus_do_not_get_bonus_rolls()
     {
-        var rolls = new[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 10, 10 };
-        var game = RollMany(rolls, new BowlingGame());
-        Assert.Equal(30, game.Score());
+        var previousRolls = new int[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 10, 10};
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        Assert.Equal(30, sut.Score());
     }
 
     [Fact(Skip = "Remove to run test")]
     public void A_strike_with_the_one_roll_bonus_after_a_spare_in_the_last_frame_does_not_get_a_bonus()
     {
-        var rolls = new[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 3, 10 };
-        var game = RollMany(rolls, new BowlingGame());
-        Assert.Equal(20, game.Score());
+        var previousRolls = new int[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 3, 10};
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        Assert.Equal(20, sut.Score());
     }
 
     [Fact(Skip = "Remove to run test")]
     public void All_strikes_is_a_perfect_game()
     {
-        var rolls = new[] { 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10 };
-        var game = RollMany(rolls, new BowlingGame());
-        Assert.Equal(300, game.Score());
+        var previousRolls = new int[] {10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10};
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        Assert.Equal(300, sut.Score());
+    }
+
+    [Fact(Skip = "Remove to run test")]
+    public void Rolls_cannot_score_negative_points()
+    {
+        var previousRolls = new int[] {};
+        var sut = new BowlingGame();
+        foreach (var roll in previousRolls)
+        {
+            sut.Roll(roll);
+        }
+        sut.Roll(-1);
+        Assert.Null(sut.Score());
```

Comment:
I would like the cases where the the canonical data indicates an error to throw an exception. This means you have to add something like this:

`canonicalDataCase.ExceptionThrown = canonicalDataCase.Expected is long ? null : typeof(ArgumentException);`

It probably also means removing the code explicitly setting the tested property to `Score` and the expected value to `null`.

Reply:
Aha, I'll update it to do this. I was following how the hand made test cases were. I'll do the update for this, it should make the code nicer.